doc/api/async_hooks.md

```diff
@@ -0,0 +1,275 @@
+
+
+// Wait for process to exit and then print the data about async operations we collected
+process.on('exit', onexit)
+function onexit() {
+  // Stop listening on the above hooks
+  asyncHook.disable()
+  for (let v of activities.values()) {
+    console.log(v)
+  }
+}
+```
+
+
+### Output
+
+
+Here we can see that a Timeout and a TIMERWRAP were created and when their lifetime events occurred.
+
+
+```
+{ uid: 2,
+  type: 'Timeout',
+  parentUid: 1,
+  init: [ '1.13 ms' ],
+  before: [ '107.72 ms', '209.41 ms' ],
+  after: [ '108.47 ms', '210.14 ms' ],
+  destroy: [ '209.91 ms', '210.17 ms' ] }
+{ uid: 3,
+  type: 'TIMERWRAP',
+  parentUid: 1,
+  init: [ '1.72 ms' ],
+  before: [ '107.31 ms', '209.23 ms' ],
+  after: [ '108.54 ms', '210.18 ms' ],
+  destroy: [ '210.20 ms' ] }
+```
+
+
+We will refer back to this example in the remaining documentation.
+
+
+## API
+
+
+### async_hooks.createHook()
+
+
+Registers functions to be called for different lifetime events of each async operation. The hooks are as follows and are
+all optional. Returns an `AsyncHook`
+
+
+* `init` {function}
+  * called _exactly_ once whenever an async operation is created
+  * arguments
+    * `uid` {number} a unique id for the async operation
+    * `type` {String} the type of the async operations, the types and their meanings are explained futher below
+      TODO(thlorenz)
+    * `triggerId` {number} the unique id of the async operation in which context this async operation was created
+    * `handle` {Object} reference to the handle representing the async operation, needs to be released during _destroy_
+* `before` {function}
+  * called whenever a callback is invoked from the context of this async operation
+  * arguments
+    * `uid` {number} a unique id for the async operation
+* `after` {function}
+  * called whenever a callback that was invoked from the context of this async operation finished executing
+  * arguments
+    * `uid` {number} a unique id for the async operation
+* `destroy` {function}
+  * called _exactly_ once whenever an async operation is destroyed
+  * arguments
+    * `uid` {number} a unique id for the async operation
+
+
+### AsyncHook.enable
+
+
+Enables any subscriptions of an `AsyncHook` instance registered via [async_hooks.createHook](async_hooks.createHook).
+
+
+### AsyncHook.disable
+
+
+Disables any subscriptions of an `AsyncHook` instance registered via [async_hooks.createHook](async_hooks.createHook).
+
+
+### async_hooks.currentId()
+
+
+Returns the uid of the current context.
+
+
+### async_hooks.triggerId()
+
+
+Returns the uid of the context that is the parent of the current context.
+
+
+## Understanding the _setTimeout_ Example
+
+
+TODO(thlorenz)
+
+
+Will explain the example step by step here. Not sure if this should be placed below the embedder API or if it's ok here.
+Reason for being here is that people who just want to use this never have to even scroll to the much more complex
+embedder API.
+
+
+## Embedder API
```

Comment:
OK will dig into that a bit more `AsyncEvent` and update the docs as far as I'll understand things at that point ...